accounting/gst.js

```diff
@@ -0,0 +1,197 @@
+  if (!savePath) return;
+
+  const gstData = {
+    version: 'GST3.0.4',
+    hash: 'hash',
+    gstin: printSettings.gstin,
+    // fp is the the MMYYYY for the last month of the report
+    // for example if you are extracting report for 1st July 2020 to 31st September 2020 then
+    // fb = 092020
+    fp: DateTime.fromISO(toDate).toFormat('MMyyyy'),
+  };
+
+  // based condition we need to triggered different methods
+  if (transferType === 'B2B') {
+    gstData.b2b = await generateB2bData(report.rows);
+  } else if (transferType === 'B2CL') {
+    gstData.b2cl = await generateB2clData(report.rows);
+  } else if (transferType === 'B2CS') {
+    gstData.b2cs = await generateB2csData(report.rows);
+  }
+
+  await sleep(1);
+  const jsonData = JSON.stringify(gstData);
+  makeJSON(jsonData, savePath);
+}
+
+async function generateB2bData(invoices) {
+  const b2b = [];
+
+  invoices.forEach(async (row) => {
+    // it's must for the customer to have a gstin, if not it should not be here
+    const customer = {
+      ctin: row.gstin,
+      inv: [],
+    };
+
+    const invRecord = {
+      inum: row.invNo,
+      idt: DateTime.fromFormat(row.invDate, 'yyyy-MM-dd').toFormat(
+        'dd-MM-yyyy'
+      ),
+      val: row.invAmt,
+      pos: row.gstin && row.gstin.substring(0, 2),
+      rchrg: row.reverseCharge,
+      inv_typ: 'R',
+      itms: [],
+    };
+
+    let items = await frappe.db
+      .knex('SalesInvoiceItem')
+      .where('parent', invRecord.inum);
+
+    items.forEach((item) => {
+      const itemRecord = {
+        num: item.itemCode || 0,
+        itm_det: {
+          txval: item.baseAmount,
+          rt: GST[item.tax],
+          csamt: 0,
+          camt: ((CSGST[item.tax] || 0) * item.baseAmount) / 100,
+          samt: ((CSGST[item.tax] || 0) * item.baseAmount) / 100,
+          iamt: ((IGST[item.tax] || 0) * item.baseAmount) / 100,
+        },
+      };
+
+      invRecord.itms.push(itemRecord);
+    });
+
+    const customerRecord = b2b.find((b) => b.ctin === row.gstin);
+
+    if (customerRecord) {
+      customerRecord.inv.push(invRecord);
+    } else {
+      customer.inv.push(invRecord);
+      b2b.push(customer);
+    }
+  });
+
+  return b2b;
+}
+
+async function generateB2clData(invoices) {
+  return [];
+}
+
+async function generateB2csData(invoices) {
+  return [];
+}
+
+async function getSavePath(name) {
+  const options = {
+    title: _('Select folder'),
+    defaultPath: `${name}.json`,
+  };
+
+  let { filePath } = await ipcRenderer.invoke(
+    IPC_ACTIONS.GET_SAVE_FILEPATH,
+    options
+  );
+
+  if (filePath) {
+    if (!filePath.endsWith('.json')) {
+      filePath = filePath + '.json';
+    }
+  }
+
+  return filePath;
+}
+
+export function promptWhenGstUnavailable() {
+  return new Promise((resolve) => {
+    showMessageDialog({
+      message: _('Export failed!'),
+      description: _(
+        'Report cannot be exported if company gst details are not configured'
+      ),
```

Comment:
The title (`message`) can be "Export Failed", no need for punctuation in the title. Also full stop at the end of `description`.

p.s. I know this is very pedantic but these are almost zero cost improvements to maintain consistency